cylc/uiserver/authorise.py

```diff
@@ -0,0 +1,501 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from contextlib import suppress
+from functools import lru_cache
+import grp
+from typing import List, Dict, Optional, Union, Any, Sequence, Set
+from inspect import iscoroutinefunction
+import logging
+import os
+from tornado import web
+from traitlets.config.loader import LazyConfigValue
+from traitlets.traitlets import Bool
+
+logger = logging.getLogger(__name__)
+
+
+class Authorization:
+    """Authorization Information Class
+    One instance for the life of the UI Server. If authorization settings
+    change they will need to re-start the UI Server.
+    Authorization has access groups: `READ`, `CONTROL`, `ALL` - along with
+    their negations, `!READ`, `!CONTROL` and `!ALL` which indicate removal of
+    the permission groups.
+
+    """
+
+    # config literals
+    DEFAULT = 'default'
+    LIMIT = 'limit'
+    GRP_IDENTIFIER = 'group:'
+
+    # Operations
+
+    ##########################################################################
+    #                               !WARNING!                                #
+    #                                                                        #
+    #   Beware of changing these permission groups. Users may be relying on  #
+    #   these settings. Changes should be widely publicised to users.        #
+    #                                                                        #
+    #   If adding/removing operations, ensure documentation is updated.      #
+    #                                                                        #
+    ##########################################################################
+
+    READ_OPERATION = "read"
+
+    # Access group identifiers (used in config)
+
+    READ = "READ"
+    CONTROL = "CONTROL"
+    ALL = "ALL"
+    NOT_READ = "!READ"
+    NOT_CONTROL = "!CONTROL"
+    NOT_ALL = "!ALL"
+
+    # Access Groups
+
+    READ_OPS = [
+        "ping",
+        "read",
+    ]
+
+    CONTROL_OPS = [
+        "extTrigger",
+        "pause",
+        "kill",
+        "message",
+        "hold",
+        "play",
+        "poll",
+        "release",
+        "releaseholdpoint",
+        "reload",
+        "remove",
+        "resume",
+        "setgraphwindowextent",
+        "setholdpoint",
+        "setoutputs",
+        "setverbosity",
+        "stop",
+        "trigger",
+    ]
+
+    ALL_OPS = [
+        "ping",
+        "read",
+        "broadcast",
+        "extTrigger",
+        "pause",
+        "kill",
+        "message",
+        "hold",
+        "play",
+        "poll",
+        "release",
+        "releaseholdpoint",
+        "reload",
+        "remove",
+        "resume",
+        "setgraphwindowextent",
+        "setholdpoint",
+        "setoutputs",
+        "setverbosity",
+        "stop",
+        "trigger",
```

Comment:
Do we have a way to generate these lists dynamically?

Reply:
I think this is a great idea. The mutations could each have the correct auth access group flag and get sent from cylc-flow. I tries this initially and got into a bit of a tangle but I think it is worth pursuing. I will put up a follow up issue for it.

Reply:
I think this should do the trick:

```python
import graphene

from cylc.uiserver.schema import UISMutations

[
    attr
    for attr in dir(UISMutations)
    if isinstance(getattr(UISMutations, attr), graphene.Field)
]
```

ATM only read mutation is `ping` so we can hardcode that part.

In the long run perhaps we could define a graphene `Meta` field e.g.

```python
class Hold(Mutation, TaskMutation):                                                  
    class Meta:
        description = sstrip('''
            Hold tasks within a workflow.
        ''')
        access = 'CONTROL'

```